### ISSUE TITLE
Rapport d'implémentation

### ISSUE THREAD

**Issue by DanhierPierre** [axtux/data-diode#3]
Dans le rapport d'implémentation il ne faut pas oublier de lier tout ce qui est sécurités avec les contre-mesures détaillées dans l'analyse de risques.

**Comment by DanhierPierre** [comment]
Et ne pas oublier non plus de détailler pourquoi on a choisi de ne pas laisser IP fixe dans les choix.

**Comment by lnyaki** [comment]
Le problème, c'est que mon rapport d'implémentation, il contient presque rien. Là où il y a des trucs, c'est le démo plan.

Parce que si je me souvient, le rapport d'implémentation, ça devait décrire le système tel qu'il serait au final, tandis que le démo plan explique comment on a vraiment implémenté le prototype (en l'occurence, avec docker, etc).

Et le système tel qu'il sera au final, ben c'est une boite avec deux serveurs et un cable de fibre optique. En gros, c'est ça. (ou alors j'ai mal compris, ce qui est possible).

**Comment by lnyaki** [comment]
Par contre, vis à vis de l'ip, on peut pas simplement mettre qu'elle est fixe? C'est la solution la plus simple, c'est la solution préconisée par le prof, et elle ne fait pas controverse.

Parce que aller avec le DHCP, on va devoir défendre ça à l'examen, et j'ai pas l'impression que ça va etre facile, surtout que personnellement je suis plutot d'accord avec le prof.

Parce que faut pas oublier que les gars ils vont nous fusiller sans aucune pitié au moindre truc qu'on ne sait pas parfaitement justifier.

**Comment by DanhierPierre** [comment]
Si je me souviens bien, le rapport d'implémentation c'est "comment on aurait fait avec le matos" . Donc  expliquer qu'on aurait installé 2 serveurs Ubuntu sur des machines , qu'il y aurait des serveurs web, php, ... Et le demo plan ça doit juste etre "On a pas de matos, on a remplacé ça par des containers et des IP tables + montrer les IP tables" Je crois

**Comment by lnyaki** [comment]
Ok, c'est bien ce que j'avais compris.

**Comment by DanhierPierre** [comment]
mais donc toute la partie logicielle dans qu'on a dans les containers, c'est dans l'implémentation qu'il faut la détailler en liant avec les contres mesures. Pour DHCP c'est délicat de dire " on a fait ça à la démo mais on a changé" :s

**Comment by lnyaki** [comment]
Oui, la parite logicielle, c'est  dans implémentation, mais c'est juste un serveur http et une db mysql. 

Pour DHCP, j'aurais pas de problème à changer. Après tout, la démo n'avait rien avoir avec la question du DHCP. Le DHCP, même si on l'a mentionné, ben c'est dans le rapport qu'on le décrirait, or, le rapport n'est pas encore rendu.

**Comment by DanhierPierre** [comment]
ok bah si tu crois que ça ne nous retombera pas dessus si on change, c'est sur que ça vaut mieux e mettre IP fixe

**Comment by DanhierPierre** [comment]
Enfin l'idéal c'est de permettre les 2

**Comment by lnyaki** [comment]
Je ne suis pas certain que ça ne nous retombera pas dessus. Mais  par contre, le DHCP, ça nous retombera dessus avec certitude.

**Comment by DanhierPierre** [comment]
C'est pas faux

**Comment by DanhierPierre** [comment]
Ce qu'il manque majoritairement dans la partie actuelle de l'implémentation c'est comment on aurait implémenté les contre mesures. Genre dire " En lien avec la contre mesure 10 et 11, on push les données 3 fois" et pour le flooding, vu qu'on sait pas gérer ça à notre niveau, on dirait pas de mettre un reverse proxy ? Enfin, est-ce qu'on ne mettrai pas une petite explication de comment joindre la db pour le client, histoire qu'il puisse développer ce qu'il veut avec après ? Au lieu de dire " démerde toi et cherche comment la contacter" ?

**Comment by lnyaki** [comment]
Donc tu veux un détailé de toutes les contre mesures possibles dans l'implém, c'est ça?

**Comment by DanhierPierre** [comment]
Bah expliquer l'implémentation de chaque contre mesures qu'on a mise en place. Donc y a le push 3 fois chaque paquets à cause d'udp, les requetes préparées pour éviter SQL injection, les tokens dans les liens des pages pour eviter CSRF, l'obligation de password un minimum longs pour eviter le bruteforce et enfin pour le flooding, comme on a rien contre ça, je pensais mettre un reverse proxy mais faut voir si c'est bien comme solution.

**Comment by lnyaki** [comment]
Le problème, c'est que là, j'ai besoin d'Alexis (je ne sais pas où il est), parce que c'est lui qui a implémenté le code niveau serveur. Je me suis occupé de l'infrastructure docker, mais tout ce qui concerne le passage et le traitement des données, c'est Alexis. Tu sais où il est?

**Comment by DanhierPierre** [comment]
Je lui ai envoyé un message pour luid ire qu'on avait besoin de lui, il m'a dit qu'il finissait un truc donc je suppose qu'il va venir après avoir fini, tu lui a envoyé un mail avec ce que tu veux qu'il fasse ?

**Comment by lnyaki** [comment]
Salut,

C'est pas tellement que j'ai besoin qu'il fasse des choses de son coté, mais j'ai besoin qu'il m'épaule un peu pour certaines choses.

Là je travaille sur les counter measures, dans la section implémentation. Et par exemple, pour le risk 8 (brute force attack pour craquer le mot de passe des utilisateurs), il est spécifié qu'on demande un long mot de passe. Seulement, dans le code, on ne fait que vérifier que le mot de passe n'est pas vide. Donc un mot de passe de 1 charactère est possible.

**admin.php**

```
function add_id(&$users, $id, $pass)
{
  if (empty($id))
  {
    error('Empty login.');
  }
  
  // do not allow empty password
  $salt = random_id();
  $hash = empty($pass) ? null : md5($salt.$pass);
  $users[$id] = $salt.'|'.$hash;
  return true;
}

```

**Comment by DanhierPierre** [comment]
On envoie pas notre implémentation donc c'est pas grave , on peut dire qu'on impose des mot de passe longs.  Au final ça prendrait que 2-3 minutes pour changer ça réellement

**Comment by lnyaki** [comment]
Oui, mais on doit mettre des morceaux de code pour illustrer comment on a fait non? Sinon, il n'y a pas de sens à mettre ces trucs là dans la section implémentation. Puisque s'il s'agit juste de dire "on force une longueur de 8 charactères et la présence d'une majuscule et d'un charactère spécial", ben c'est ce qu'on dit dans la section 7.6 counter measures (en l'occurence, on a pas spécifié la taille des mots de passe là, mais je vais le rajouter).

Et donc, si on met pas de code, y a rien à mettre dans la section implem alors. (Tu comprends ce que je veux dire?).

**Comment by DanhierPierre** [comment]
Oui, bah faut voir si Alexis sait changer ça vite fait et on met le code, si pas on n'en met juste pas pour cet exemple.

**Comment by lnyaki** [comment]
Oui, mais tu vois, je pense que pour les countre mesures des risques 6 et 7 aussi, on a en fait pas d'implémentation non plus. Donc je ne sais rien mettre dans la section implementation pour ces contre mesures là.

**Comment by DanhierPierre** [comment]
L'implémentation c'est "comment on aurait fait avec le budget, le matos et le temps nécessaire" donc on peut détailler comment on aurait fait, pas forcément besoin de montrer une implémentation. Je crois donc qu'une simple explication de comment on aurait implémenté ces contre mesures suffis

**Comment by lnyaki** [comment]
Ah, je comprends mieux, ok. 

Mais alors quelle est la différence entre la section 7.6 countermeasures, et la section implémentation?

Parce que dans la 7.6 on explique déjà les counter measures. Donc dans la section implementation, on ne va faire que se répéter?

Par exemple dans le _7.67 , risk 7 : flooding of the outside webpage_,  on dit : we mitigate risk by blocking opponent's ip address [...]

Mais alors, dans la section implémentation, qu'est-ce qu'on dit? On redis la meme chose? Ou on essaie d'ajouter des détails alors?

**Comment by DanhierPierre** [comment]
Bah tu explique concrètement comment on implémenterai ça, est ce qu'on ajoute cette vérification dans un logiciel, lequel, .. ou alors on met juste un reverse proxy et on explique ce que ça fait et pourquoi ça remplit son rôle . Perso je vois la partie contre mesure comme "en gros on va faire ça pour diminuer ou eviter le risque" et la partie implem comme " Concrètement, on le fait ou on l'aurait fait comme ça " mais pas forcément besoin de montrer un code. L'idéal c'est de montrer le code, mais on en a pas pour tout ^^

**Comment by lnyaki** [comment]
Ok, je pense avoir pigé. Je vais faire de mon mieux.

**Comment by lnyaki** [comment]
par contre, est ce que tu aurais le temps de faire un schema du systeme, avec les 3 réseaux représentés, et les adresses ip des servers spécifiés?

**Comment by DanhierPierre** [comment]
Ouais je peux faire ça , tu sais m'envoyer un résumé des adresses et ports ?

**Comment by lnyaki** [comment]
Biensur.

Je résume (les détails sont en dessous): (si tu comprends pas, dis moi)

**Réseaux**

- 200.0.0.0/16 outside-network
- 10.0.0.0/24 secure-network
- 192.168.0.0/24 intra-diode

**Servers**
-  200.0.0.2 --[sender]---  192.168.0.3  ----------------->  192.168.0.2 --[receiver]-- 10.0.0.2

**Ports**
200.0.0.2 : http 80 (bindé sur 8881 sur le host) et https 443 (bindé sur 8882)
10.0.0.2  : http 80 (bindé sur le 8880 du host)
192.168.0.2: udp 4242

Pour les réseaux, je te met l'extrait du script de creation des réseaux : 

```
sudo docker network create --subnet=200.0.0.0/16 outside-network
sudo docker network create --subnet=10.0.0.0/24 secure-network
sudo docker network create --subnet=192.168.0.0/24 intra-diode
```

Pour les addresses, c'est  (extrait du fichier de création de containers)

```
# Building container : inner side of the diode
cd internal/
sudo docker build . -t diode_in-img

sudo docker run -t -d --net secure-network --ip 10.0.0.2 -p 8880:80 --name diode-in --cap-add=NET_ADMIN --cap-add=NET_RAW diode_in-img
cd -

# Building container : outter side of the diode
cd external/
sudo docker build . -t diode_out-img 
sudo docker run -t -d --net intra-diode --ip 192.168.0.3 -p 8881:80 -p 8882:443 --name diode-out diode_out-img
cd -

# Creating connecting a part of the diode to the outside, and the other one to the inside
sudo docker network connect --ip 192.168.0.2  intra-diode diode-in
sudo docker network connect --ip 200.0.0.2 external-network diode-out
```

**Comment by lnyaki** [comment]
Voilà, toutes les infos sont là (j'ai du remettre le post précédent à jour plusieurs fois).
Là je vais manger un bout, je reviens dans 20 min (enfin, je reste devant mon pc, mais je suis en train de manger).

**Comment by DanhierPierre** [comment]
ok, la page web interne , on y accede comment du coup ? on avait pas exposé un port ?

**Comment by DanhierPierre** [comment]
J'ai push une première version, regarde si je me trompe pas, et si tu veux que je l'améliore dis moi. img/reseau.png

**Comment by lnyaki** [comment]
Salut, pour la page web interne, on y accède via le browser du host, car on fait un bind du port 80 de la partie interne de la diode vers le port 8880 du host (comme tu l'as correctement mis dans ton schéma.

Sinon, ben ça a l'air d'aller pour le schéma.

**Comment by DanhierPierre** [comment]
Okok, si t'as encore besoin de moi hésite pas

**Comment by lnyaki** [comment]
Est ce que tu peux regarder comment importer les svg en latex? J'essaie mais je n'y arrive pas, ça me fait des erreurs à la compilation.

Je me base sur  ça : https://tex.stackexchange.com/questions/2099/how-to-include-svg-diagrams-in-latex
et ça :  http://blog.chapagain.com.np/latex-include-and-scale-svg-scalable-vector-graphics-image/

**Comment by DanhierPierre** [comment]
J'ai mis un .png aussi le .svg c'est juste si on veut modifier l'image

**Comment by lnyaki** [comment]
Je sais, mais le png est tout pixelisé, alors que le svg est tout lisse.

**Comment by DanhierPierre** [comment]
J'ai push une version un peu dépixelisée mais je sais pas faire mieux et je sais comment mettre un .svg dans latex :( Par contre Alexis vient de me dire qu'il savait pas être la .. on doit se démerder pour rendre un truc à 2...

**Comment by lnyaki** [comment]
Pour Alexis, c'est ok. Y a pas beaucoup de code, donc je sais vite trouver ce don't j'ai besoin.

Pour les png, ben tant pis alors.

**Comment by DanhierPierre** [comment]
ça rend si mal que ça ? Je préfère mettre une image pixélisée où on comprend que rien . Et faut plutot se focus sur le contenu.

**Comment by lnyaki** [comment]
T'inquiète, je suis focalisé sur le contenu, mais c'est pour ça que je te le demandais. Si t'avais le temps.

Là j'ai presque fini. Je pusherai vers 9h. Ensuite on prendra 30 minutes pour relire, et 30 minutes pour les dernières retouches.

Donc on sera prèt aux alentours de 22h. Ce qui nous laisse 2h de marge en cas de problème.

Ca te conviens?

Loïc.

**Comment by DanhierPierre** [comment]
Pas de soucis, je suis la

**Comment by lnyaki** [comment]
Je viens de push.

Il me reste à spécifier l'adresse IP par défaut dans la section installation. Mis à part ça, c'est plus ou moins tout.

Est ce que tu peux commencer à relire et me dire s'il y a encore un soucis quelque part? J'atttends ton feeback entre 21h30 et 21h45. Mais si c'est un peu après, c'est pas trop grave.

**Comment by DanhierPierre** [comment]
OK, je commence à relire

**Comment by DanhierPierre** [comment]
Alors, pour le feedback , faut mettre en page de manière à ce que des images ou des parapgraphes ne soient pas séparés entre 2 pages;
Dans la description de malicious user tu dis que c'est un employé alors que c'ets plutot un mec avec un appareil connecté qui envoie de données en trafiquant son appareil ou en chopant les identifiants dedans et en les utilisants sur un pc
Plutôt parler de sessions plutôt que de requêtes dans le 8.2.6 ?
Et enfin, expliquer comment joindre la db une fois le système en place nan ?
Sinon pour moi c'est ok. Il ne reste que des petits détails quoi .

**Comment by lnyaki** [comment]
Pour moi, ça a l'air à peu près bon, mis à part que la fin m'a l'air un peu abrute (pas de conclusion, pas de dernier mot pour résumer le projet. Je pense que je vais en rajouter. Qu'en penses tu?).

Je vais commencer à faire une espèce de conclusion, on vera ce que ça donne.

**Comment by DanhierPierre** [comment]
Une conclusion c'est toujours bien mais je vois pas trop quoi dire perso

**Comment by lnyaki** [comment]
Pour la conclusion, je suis en train de la faire. Je te la donnerai à lire, et tu pourra me dire si ça a du sens ou non. Je trouve que ça pourrait être pas mal.

Concernant tes remarques sur la mise en page, ben parfois latex place certains éléments lui même, et on sait pas controller.

Pour malicious user : ok, j'avais mal compris. Je suis d'accord avec ta remarque.

Pour la question sur le CSRF : le token n'est pas fixe pour une session (il est stocké dans la session, mais change à chaque fois que l'utilisateur va sur une nouvelle page (à chaque requete))

Par contre, je remarque qu'on ne mentionne pas du tout les objets connectés et tout ça, les hopitaux, etc. Mais bon, c'est peut être mieux comme ça, parce que ça ouvrait à trop de pièges.

**Comment by DanhierPierre** [comment]
Pour les appareils connectés on en parle quand on explique l'api . C'est vrai qu'on en aprle pas beaucoup mais y a pas vraiment grand chose à dire . Ce qui me gène plus c'est qu'on ne parle pas de comment utiliser la db pour le client .

**Comment by lnyaki** [comment]
Je ne comprend pas ce que tu veux dire. Le client, il ne touche pas à la db.
Là db se lance quand la data-diode se lance. Les données y sont stockées automatiquement, et le client peut consulter ces données à travers l'interface web.

Il n'a pas d'autre role vis à vis de la db.

**Comment by lnyaki** [comment]
J'ai écris une conclusion. Qu'en penses-tu? Ca passes?

**Comment by DanhierPierre** [comment]
Pour la db, le client doit pouvoir y accéder pour l'utiliser dans ses programmes, c’est ça que je veux dire.

**Comment by DanhierPierre** [comment]
pour la conclusion, y a O majuscule dans Concept à retirer x) Et au final on a pas une data diode, mais une simulation de data diode ( on a pas réellement un appareil avec une liaison unidirectionnelle physique et l'autre physiquement coupée ) . Sinon c'est ok ;)

**Comment by lnyaki** [comment]
Pour tes remarques sur la conclusion : ok, je suis d'accord.

Pour la db : j'ai pas l'impression qu'on ai prévu qu'il puisse se connecter lui même (même si je sais qu'on l'a mentionné), donc je ne sais pas ce que je suis sensé mettre à ce sujet.

**Comment by DanhierPierre** [comment]
Bah une fois qu'il a les données dans le db, faut bien qu'il les utilisent ou les affiche quoi x) C'est ce que je veux dire

**Comment by lnyaki** [comment]
Ils les affichent en se connectant en http sur la data-diode (coté interne).

**Comment by DanhierPierre** [comment]
Oui mais ça c'est juste un moyen qu'on a  mis en place pour visualiser les données pour la démo en attendant que le client en fasse quelque chose. Juste les afficher ce n'est pas super utile , le but de notre système c'est que le client prenne ces données , fasse de l'analyse avec en les utilisant

**Comment by DanhierPierre** [comment]
Donc il faudrait juste expliquer dans le rapport comment il pourrait y accéder , mais si c'est trop tard tant pis, on peut envoyer pour moi

**Comment by lnyaki** [comment]
D'accord, mais on a rien de ça implémenté, et du coup je ne sais pas ce que tu veux que je mette.

**Comment by lnyaki** [comment]
Je vais essayer de rajouter un truc. Si t'as une idée de quoi mettre, dis moi.

**Comment by DanhierPierre** [comment]
bah en gros faut juste montrer comment notre page web accède aux données et dire que le client peut faire pareil

**Comment by lnyaki** [comment]
C'est quoi le nom des profs? L'assistant c'est thibault debatty. Le prof, tu sais son nom? C'est pour rajouter sur la page de garde.

Concernant la db, je rajoue ça où? Dans implémentation?

**Comment by DanhierPierre** [comment]
oui dans l'implémentation c'est bon je crois
Le nom du prof c'est MEES Wim

**Comment by lnyaki** [comment]
Pour la db, ça va pas. Parce qu'en fait, il faudrait passer à l'utilisateur le nom d'utilisateur et mot de passe de la db (comment, par quel processus). Et après, si quelqu'un vient à modifier le mot de passe de la db (en modifiant un script), alors comment être certains que les utilisateurs aient connaissance de ce nouveau mot de passe?

On peut mettre des mots de passe configurables, mais alors il faut rajouter des nouveaux risques et tout.

Avec le temps qu'il reste, je ne suis pas certain qu'on puisse considérer ces nouveaux cas sans faire de bêtise.

**Comment by DanhierPierre** [comment]
OK bah tant pis, on a plus le temps.

**Comment by lnyaki** [comment]
Là j'essaie de mettre le nom des prof en bas de la page de garde. Tu sais comment faire?

**Comment by lnyaki** [comment]
Ah, j'ai trouvé (\vfill)

**Comment by DanhierPierre** [comment]
ah j'avais trouvé une autre façon avec ce tuto http://ww2.ac-poitiers.fr/math/spip.php?article267 mais si t'as trouvé c'est nickel ^^

**Comment by lnyaki** [comment]
Voillà, j'ai commité la dernière version. Le nom du prof est un peu décallé, mais tant pis, j'ai pas trouvé le problème.

J'envoie?

**Comment by DanhierPierre** [comment]
Perso j'ai pas  le nom des profs, t'es sur d'avoir push?

**Comment by lnyaki** [comment]
Essaie de re-puller? Parce que moi j'ai pushé.

**Comment by DanhierPierre** [comment]
C'est bizarre, dans ton commit je vois la différence mais quand je pull sur ma vm et sur mon mac je les ai pas

**Comment by DanhierPierre** [comment]
En downloadant direct la ça marche, tu peux envoyer ;) Oublie pas de renommer le fichier ;)

**Comment by lnyaki** [comment]
Ca roule.

**Comment by lnyaki** [comment]
Ok, c'est envoyé.

Merci pour ton aide. J'espère que ça passera.

**Comment by DanhierPierre** [comment]
J'espère aussi !
